File: exercises/cookbook/logging.cpp
// Appendix F, Recipes 15 and 24 - print a diagnostic you will actually see;
// compile a diagnostic out of Release.
//
// The three functions below are quoted VERBATIM in book/F-rosetta-cookbook.md:
// editing one means editing the appendix in the same commit (the testlab
// discipline). main() is scaffolding - it captures cerr into a buffer to
// assert the text, so a green run stays silent on stderr. build_all.sh
// builds this TU twice, the second time with -DNDEBUG, because Recipe 24's
// claim is about what vanishes under that define.
#include <cassert>
#include <iostream>
#include <sstream>
#include <string>

void report_progress(int done, int total) {
    std::cout << "processed " << done << " of " << total << '\n';    // buffered: fast
}

void report_failure(const std::string& what) {
    std::cerr << "error: " << what << '\n';    // unbuffered: survives a crash
}

// Recipe 24 - [Conditional("DEBUG")]
void check_channel_count([[maybe_unused]] int channels) {          // used only in Debug
    assert(channels > 0 && "a session has at least one channel");   // gone under NDEBUG
#ifndef NDEBUG
    std::cerr << "[debug] channels=" << channels << '\n';           // and so is this block
#endif
}

int main() {
    std::ostringstream captured;
    std::streambuf* old = std::cerr.rdbuf(captured.rdbuf());
    report_failure("cannot open settings.ini");
    std::cerr.rdbuf(old);
    assert(captured.str() == "error: cannot open settings.ini\n");

    report_progress(3, 10);    // stdout; build_all.sh silences it

    // Recipe 24: both spellings compile to nothing under NDEBUG. Plain ifs,
    // because assert itself is the thing under test.
    std::ostringstream debug_lines;
    old = std::cerr.rdbuf(debug_lines.rdbuf());
    check_channel_count(2);
    std::cerr.rdbuf(old);
    int bumps = 0;
    assert(++bumps == 1);      // the trap: a side effect inside an assert
#ifdef NDEBUG
    if (!debug_lines.str().empty() || bumps != 0) return 1;   // both vanished with the define
#else
    if (debug_lines.str() != "[debug] channels=2\n" || bumps != 1) return 1;
#endif
    return 0;
}

File: exercises/cookbook/rust/src/logging.rs
//! Appendix F, Recipes 15 and 24 - print a diagnostic you will see; compile
//! one out of Release. Included by book/F-rosetta-cookbook.md between the
//! recipe-N markers: edit here and the page follows. The tests are
//! scaffolding and appear in no listing.

// --8<-- [start:recipe-15]
pub fn report_progress(done: usize, total: usize) {
    println!("processed {done} of {total}");    // stdout: line-buffered, and a terminal shows it
}

pub fn report_failure(what: &str) {
    eprintln!("error: {what}");    // stderr: unbuffered, survives a crash, separable in a pipe
}
// --8<-- [end:recipe-15]

// --8<-- [start:recipe-24]
pub fn check_channel_count(channels: i32) {
    debug_assert!(channels > 0, "a session has at least one channel");    // gone in --release
    if cfg!(debug_assertions) {
        eprintln!("[debug] channels={channels}");    // and so is this block: cfg! is a constant
    }
}
// --8<-- [end:recipe-24]

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn diagnostics_print_without_panicking() {
        report_progress(3, 10);
        report_failure("disk full");
        check_channel_count(2);
    }

    #[test]
    #[cfg(debug_assertions)]    // the assertion exists only where debug_assert! does
    #[should_panic(expected = "at least one channel")]
    fn zero_channels_is_refused_in_debug() {
        check_channel_count(0);
    }
}
